### ISSUE TITLE
[BUG] If using useInterval with setState, and setState returns the same value, memory leaks

### ISSUE THREAD

**Issue by arturcarvalho** [juliencrn/usehooks-ts#592]
### Describe the bug

The unexpected behaviour is the memory going up when the value set with useState is always the same (in this case when `if(prev <=0)` is true and starts always setting 0. Code used:

```ts
function App() {
  const [charge, setCharge] = useState<number>(1000);

  function updateRemainingCharge() {
    setCharge((prev) => {
      if (prev <= 0) return 0;
      return prev - 1;
    });
  }

  useInterval(updateRemainingCharge, 5);

  return (
    <div>
      <h1>{charge}</h1>
    </div>
  );
}
```

What I tried up to the moment:
- Check development and build modes (leaks)
- Build with vite and nextjs with default settings (leaks)
- Use useMemo (leaks)

Any help would be greatly appreciated.
Artur

### To Reproduce

To reproduce locally: 
build it from the repo [https://github.com/arturcarvalho/ovp](https://github.com/arturcarvalho/ovp)

To reproduce it deployed:
Visit [https://ovp-two.vercel.app/](https://ovp-two.vercel.app/)

I've checked the memory leak on chrome incognito on a mac and using the developer tools 

### Expected behavior

The memory does not go up when the setState tries to change it to 0.

**Comment by longnt80** [comment]
The reason your memory keeps increasing is because you don't stop the interval when the state reaches 0. Pass `null` for the second argument for `useInterval` to stop it.

```
useInterval(updateRemainingCharge, charge > 0 ? 5 : null);
```

**Comment by arturcarvalho** [comment]
Thank you very much @longnt80 ! That is indeed fixing it, and it wasn't an issue of the `usehooks-ts` library. 

Feel free to close the issue, but still I don't understand why it's not possible to set the same value again. I also saw your explanation on reddit but it's not clear to me.

Shouldn't this be valid code? Shouldn't we be able to set the same value again?

I changed the example to this and it leaks too, shouldn't this just work? And shouldn't the optimization be apply useMemo to charge?

```ts
  const [charge, setCharge] = useState<number>(0);

  function updateRemainingCharge() {
    setCharge((prev) => {
      const sameValue = maybeTrue(0.99);
      return sameValue ? prev : prev + 1;
    });
  }

  useInterval(updateRemainingCharge, 5);
```

I updated the site+repo with this new example. https://ovp-two.vercel.app/ + https://github.com/arturcarvalho/ovp

Anyway, my problem is solved thanks to you. I really appreciated your help.

Have a nice day,
Artur
